### PR TITLE
Build-A-Sentry

### PR DESCRIPTION
**Pull request by hyper2snyper**
## About The Pull Request

New Attachment! Build-A-Sentry!
Simply attach it to the rail of a gun and BOOM, the gun can be deployed as a sentry!

## Why It's Good For The Game

Alrighty so this is a new rail slot attachment.
It costs 30 points in the engi vendor and you can get 1 for 25 points in req.

What it does:
You attach it to the rail of almost any gun. The gun now can be deployed as a sentry and will have all regular sentry functions.
It will acquire its IFF signal from your ID so if you are SOM it will not target SOM.
Its firing capabilities are that of the gun you attached it too. If the gun you attached it to does not have IFF, the Sentries bullets will not have IFF so beware.
If the gun requires Unique action to cycle. The Sentry will do it.
For example, a pump shotgun will fire, then cock then fire again and so on.
After reloading the sentry, if the gun needs to be cocked. It will. For example a Martini henry sentry will fire, open chamber, and on reload close chamber.

Since the build-a-sentry is an attachment, you can fire the gun in hand with it attached.

Unique actions can be done manually by right clicking the sentry.

New control scheme for sentries and deployable guns:
Alt click will now unload the gun of a sentry/deployed gun.
Ctrl click with remove the batter of a sentry.
Right click with do the unique action of a sentry/deployed gun.

For non deployed guns/sentries:
Alt-Right Click will remove the battery of the sentry.
Ctrl Click will now deploy the gun.

(Dont worry, I put the controls on examine so James will hopefully not get twenty mentorhelps a game about it)

### Balance:
You can stick this on to basically any gun. This should scare marines because sentries don't take into account friendlies in front of them.
Past that the primary method of balance will most likely be limiting what can attach the item.

### Videos
https://media.discordapp.net/attachments/498569646014857221/884867500444114964/laser.mp4
https://media.discordapp.net/attachments/498569646014857221/884863899441180702/thermo.mp4
https://media.discordapp.net/attachments/500801682058379265/882795351822725180/ocelot.mp4
https://media.discordapp.net/attachments/500801682058379265/883473413644120084/hans.mp4
https://media.discordapp.net/attachments/631277999655747625/883131068918689872/i_solve_problems.mp4

### Images
image
image
image
image
image
image

### Things that came along:
I removed an unused Bitfield flag because I wanted its slot.
I made the Anti-mat sniper able to remove the scope so that it could fit the attachment on.

### todo: 
~Make lasguns eject/give marines ability to remove the cells.~
~Sentry batteries should be easy to swap when a lasgun sentry is deployed.~
probably make sprites better i dunno.
~make lasgun sentries ignore windows.~
~consider different deployment macros. Alt-Right-Click seems cumbersome.~

## Changelog
:cl:
add: New Attachment, the Build-A-Sentry. Available in the Engi vendor for 30 points a piece, or 1 for 25 points in Req.
add: Simply attach it to the gun and boom, it can be deployed.
add: Changed the controls for deployable guns and sentries. You can see the controls by examining the guns that can be deployed.
balance: Anti-mat sniper can now have its scope removed.
code: Removed an old bitfield that was unused in gun code.
code: Adds a ton of infrastructure for new sentry types.
/:cl:

### CODE REVIEW

**Comment by neohenry** [comment]
image

**Comment by hyper2snyper** [comment]
I solve problems.

**Comment by ilikesaying** [comment]
mfw sniper rifle sentries

**Comment by MereWalker** [comment]
Oh lord.... Sadar sentry... Will it fuck over the marines or do its job on xenos?

**Comment by Oscari-pog** [comment]
Oh no

**Comment by bowlofcereal** [comment]
_slaps roof of build-a-sentry attachment_

this bad boy can fit many anti-material sniper rifles

**Review by SplinterGP — CHANGES_REQUESTED** [review]
Way too cheap, this is unbalanced as hell

File: code/modules/projectiles/guns/flamer.dm [hunk 1]
```diff
@@ -516,13 +521,14 @@
...
-	if(gun_user.skills.getRating("firearms") < 0)
-		switch(windup_checked)
-			if(WEAPON_WINDUP_NOT_CHECKED)
-				INVOKE_ASYNC(src, .proc/do_windup)
-				return
-			if(WEAPON_WINDUP_CHECKING)
-				return
+	if(gun_user)
+		if(gun_user.skills.getRating("firearms") < 0)
+			switch(windup_checked)
```
Comment (TiviPlus) [inline code/modules/projectiles/guns/flamer.dm:526]: ?.

File: code/modules/projectiles/mounted.dm [hunk 2]
```diff
@@ -9,13 +9,20 @@
...
+	if(istype(gun, /obj/item/weapon/gun/energy))
+		var/obj/item/weapon/gun/energy/energy_gun = gun
+		has_mag = energy_gun.cell
+	else
+		has_mag = gun.current_mag
+	if(!has_mag)
 		icon_state = default_icon_state + "_e"
 	else
 		icon_state = default_icon_state
+
```
Comment (TiviPlus) [inline code/modules/projectiles/mounted.dm:25]: Why change this still looks like icon state only updates to me

File: code/modules/projectiles/sentries.dm [hunk 3]
```diff
@@ -440,3 +454,25 @@
...
+	if(!istype(internal_item, /obj/item/weapon/gun/revolver))
+		return
+	to_chat(user, span_notice("It is as if he were still with us.")) //I miss ye already Ocelot.
+
+/obj/machinery/deployable/mounted/sentry/buildasentry/Initialize(mapload, _internal_item, deployer) //I know the istype spam is a bit much, but I don't think there is a better way.
+	. = ..()
+	name = "Deployed " + internal_item.name
+	icon = 'icons/Marine/sentry.dmi'
+	default_icon_state = "build_a_sentry"
+	update_icon()
```
Comment (TiviPlus) [inline code/modules/projectiles/sentries.dm:472]: Initialized over examine

File: code/modules/projectiles/sentries.dm [hunk 4]
```diff
@@ -440,3 +454,25 @@
 
 		distance = buffer_distance
 		return nearby_target
+
+
+/obj/machinery/deployable/mounted/sentry/buildasentry
+
+/obj/machinery/deployable/mounted/sentry/buildasentry/examine(mob/user)
```
Comment (TiviPlus) [inline code/modules/projectiles/sentries.dm:461]: you can fill the void here with something like
name = "BROKEN SENTRY"
descc = "report this on github"

File: code/modules/projectiles/sentries.dm [hunk 5]
```diff
@@ -440,3 +454,25 @@
...
+	name = "Deployed " + internal_item.name
+	icon = 'icons/Marine/sentry.dmi'
+	default_icon_state = "build_a_sentry"
+	update_icon()
+
+/obj/machinery/deployable/mounted/sentry/buildasentry/update_overlays()
+	. = ..()
+	var/obj/item/weapon/gun/internal_gun = internal_item
+	overlays.Cut()
+	overlays += image('icons/Marine/sentry.dmi', src, internal_gun.placed_overlay_iconstate, dir = src.dir)
```
Comment (TiviPlus) [inline code/modules/projectiles/sentries.dm:478]:
```suggestion
	var/obj/item/weapon/gun/internal_gun = internal_item
	. += image('icons/Marine/sentry.dmi', src, internal_gun.placed_overlay_iconstate, dir = dir)
```

**Review by TiviPlus — CHANGES_REQUESTED** [review]

File: code/modules/projectiles/mounted.dm (hunk 2 above)
Comment (BraveMole) [reply code/modules/projectiles/mounted.dm:25]: Still update icon state yes

File: code/modules/projectiles/sentries.dm [hunk 6]
```diff
@@ -91,15 +92,16 @@
 
 /obj/machinery/deployable/mounted/sentry/attackby(obj/item/I, mob/user, params)
 	var/obj/item/weapon/gun/gun = internal_item
-	if(istype(I, gun.sentry_battery_type))
-		internal_item.attackby(I, user, params)
+	if(istype(I, gun.sentry_battery_type) && !istype(gun, /obj/item/weapon/gun/energy))
```
Comment (BraveMole) [inline code/modules/projectiles/sentries.dm:95]: so how do you reload a sentry cell when the gun is an energy gun?

**Review by BraveMole — CHANGES_REQUESTED** [review]

File: code/modules/projectiles/guns/energy.dm [hunk 7]
```diff
@@ -276,6 +278,7 @@
 /obj/item/weapon/gun/energy/lasgun/update_icon(mob/user)
 	var/cell_charge = (!cell || cell.charge <= 0) ? 0 : CEILING((cell.charge / max(cell.maxcharge, 1)) * 100, 25)
 	icon_state = "[base_gun_icon]_[cell_charge]"
+	update_overlays()
```
Comment (TiviPlus) [inline code/modules/projectiles/guns/energy.dm:281]: update icon

File: code/modules/projectiles/mounted.dm [hunk 8]
```diff
@@ -27,6 +34,18 @@
...
 
+/obj/machinery/deployable/mounted/AltClick(mob/user)
+	. = ..()
+	var/obj/item/weapon/gun/internal_gun = internal_item
+	internal_gun.unload(user)
+
+/obj/machinery/deployable/mounted/attack_hand_alternate(mob/living/user)
+	. = ..()
+	if(!Adjacent(user))	
+		return
```
Comment (TiviPlus) [inline code/modules/projectiles/mounted.dm:45]: this check is already in attack hand alternate

File: code/modules/projectiles/mounted.dm [hunk 9]
```diff
@@ -40,16 +59,20 @@
 
 	reload(user, I)
 
+///Reloades the internal_item
 /obj/machinery/deployable/mounted/proc/reload(mob/user, ammo_magazine)
```
Comment (TiviPlus) [inline code/modules/projectiles/mounted.dm:63]:
```suggestion
///Reloads the internal_item
/obj/machinery/deployable/mounted/proc/reload(mob/user, ammo_magazine)
```

File: code/modules/projectiles/sentries.dm [hunk 10]
```diff
@@ -62,7 +62,7 @@
 	GLOB.marine_turrets += src
 	set_on(TRUE)
 
-/obj/machinery/deployable/mounted/sentry/update_icon_state()
+/obj/machinery/deployable/mounted/sentry/update_icon()
 	. = ..()
 	if(!CHECK_BITFIELD(machine_stat, KNOCKED_DOWN))
 		return
```
Comment (TiviPlus) [inline code/modules/projectiles/sentries.dm:68]: should be update icon state

**Review by TiviPlus — CHANGES_REQUESTED** [review]

File: code/modules/reqs/supplypacks.dm [hunk 11]
```diff
@@ -161,6 +161,15 @@ WEAPONS
 	contains = list(/obj/item/storage/box/minisentry)
 	cost = 40
 
+/datum/supply_packs/weapons/buildasentry
+	name = "Build-A-Sentry Two-Pack"
+	contains = list(
+		/obj/item/attachable/buildasentry,
+		/obj/item/attachable/buildasentry,
+	)
+	cost = 10
```
Comment (SplinterGP) [inline code/modules/reqs/supplypacks.dm:170]: still 10 points.

**Review by SplinterGP — CHANGES_REQUESTED** [review]

**Review by BraveMole — APPROVED** [review]
One last TM and after that you'll have to do the balancing

**Comment by halitosisman** [comment]
BuildaSentries

**Comment by lbnesquik** [comment]
Great PR

**Pull request merged**